funcbench/main.go

```diff
@@ -149,82 +185,78 @@ The default is 2 hours (2h).`).Default("2h").Duration()
 	logger.Println("exiting")
 }
 
-func funcbench(
+// startBenchmark returns the comparision results.
+// 1. If target is same as current ref, run sub-benchmarks and return instead (TODO).
+// 2. Execute benchmark against packages in the current worktree.
+// 3. Cleanup of worktree in case funcbench was run previously and checkout target worktree.
+// 4. Execute benchmark against packages in the new(target) worktree.
+// 5. Return compared results.
+func startBenchmark(
 	ctx context.Context,
 	env Environment,
 	bench *Benchmarker,
-) error {
+) ([]BenchCmp, error) {
+	worktreeDirName := "_funchbench-cmp"
+
 	wt, _ := env.Repo().Worktree()
 	ref, err := env.Repo().Head()
 	if err != nil {
-		return errors.Wrap(err, "get head")
+		return nil, errors.Wrap(err, "get head")
 	}
 
 	if _, err := bench.c.exec("bash", "-c", "git update-index -q --ignore-submodules --refresh && git diff-files --quiet --ignore-submodules --"); err != nil {
-		return errors.Wrap(err, "not clean worktree")
-	}
-
-	// 1. Execute benchmark against packages in the current directory.
-	newResult, err := bench.execBenchmark(wt.Filesystem.Root(), ref.Hash())
-	if err != nil {
-		// TODO(bwplotka): Just defer posting all errors?
-		if pErr := env.PostErr("Go bench test for this pull request failed"); pErr != nil {
-			return errors.Errorf("error: %v occured while processing error: %v", pErr, err)
-		}
-		return errors.Wrap(err, "exec benchmark A")
+		return nil, errors.Wrap(err, "not clean worktree")
 	}
 
-	var oldResult string
-
-	targetCommit, compareWithItself, err := compareTargetRef(ctx, env.Repo(), env.CompareTarget())
+	// Get info about target.
+	targetCommit, compareWithItself, err := getTargetInfo(ctx, env.Repo(), env.CompareTarget())
 	if err != nil {
-		return errors.Wrap(err, "compareTargetRef")
+		return nil, errors.Wrap(err, "getTargetInfo")
 	}
+	bench.logger.Println("Target:", targetCommit.String(), "Current Ref:", ref.Hash().String())
 
 	if compareWithItself {
-		bench.logger.Println("Target:", env.CompareTarget(), "is `.`; Assuming sub-benchmarks comparison.")
+		bench.logger.Println("Assuming sub-benchmarks comparison.")
+		subResult, err := bench.execBenchmark(wt.Filesystem.Root(), ref.Hash())
+		if err != nil {
+			return nil, errors.Wrap(err, "execute sub-benchmark")
+		}
 
-		// 2a. Compare sub benchmarks. TODO.
-		cmps, err := bench.compareSubBenchmarks(newResult)
+		cmps, err := bench.compareSubBenchmarks(subResult)
```

Comment:
Is this necessary ? What are sub-benchmarks ? I think we run all the sub-benchmarks with `go test -bench=. ./...`.

Reply:
sub-benchmark comparison is currently not implemented, what I assumed @bwplotka wanted to with sub-benchmarks was that you'd write benchmarks with two sub-benchmarks and they'll be compared. not sure though.

https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks

Reply:
`t.Run`

Reply:
It's not but would be nice

Reply:
AFAIK, If we specify the sub-benchmarks via regex, it will be executed within `execBenchmark`. What will `compareSubBenchmarks` do ?

Reply:
Choosing is not a problem. Comparing between subbenchmarks is. And I mean literally:

* Run now once on the current branch.
* /sub1 and /sub2 was ran.
* Compare sub1 vs sub2.

Reply:
Oh, you mean comparing with different sub-benchmarks in the same run ?
e.g.
In Run A, we executed sub1, sub2; In Run B, we executed sub1, sub2 as well.
`compareSubBenchmarks` will compare A.sub1 to A.sub2.
`compareBenchmarks` will compare A.sub1 with B.sub1, and A.sub2 with B.sub2 ?